File: engine-jsc/src/main/cpp/JSObject.cpp
#include <JavaScriptCore/JavaScript.h>
#include "JSFunction.h"
#include "JSUtils.h"


JSC_NATIVE_METHOD(JSObject,jlong,NJSmake) (JSC_NATIVE_PARAMS, jlong ctx, jlong data) {
    JSObjectRef value = JSObjectMake((JSContextRef)ctx, (JSClassRef) NULL, (void*)data);
    JSValueProtect((JSContextRef) ctx, value);
    return (long)value;
}

JSC_NATIVE_METHOD(JSObject,jlong,NJSmakeInstance) (JSC_NATIVE_PARAMS, jlong ctx) {
    JSInstance *instance = new JSInstance(env, thiz, (JSContextRef)ctx);
    return instance->getObjRef();
}

JSC_NATIVE_METHOD(JSObject,jlong,NJSmakeFunctionWithCallback) (JSC_NATIVE_PARAMS, jlong ctx, jlong name) {
    JSFunction *function = new JSFunction(env, thiz, (JSContextRef)ctx,
                                          (JSStringRef)name);
    return function->getObjRef();
}

JSC_NATIVE_METHOD(JSObject,jobject,NJSmakeArray) (JSC_NATIVE_PARAMS, jlong ctx, jlongArray args) {
    JSValueRef exception = NULL;

    int i;
    jsize len = env->GetArrayLength(args);
    jlong *values = env->GetLongArrayElements(args, 0);
    JSValueRef* elements = new JSValueRef[len];
    for (i=0; i<len; i++) {
        elements[i] = (JSValueRef) values[i];
    }
    env->ReleaseLongArrayElements(args, values, 0);

    jclass ret = env->FindClass("io/nodekit/engine/javascriptcore/JSValue$JNIReturnObject");
    jmethodID cid = env->GetMethodID(ret,"<init>","()V");
    jobject out = env->NewObject(ret, cid);

    jfieldID fid = env->GetFieldID(ret , "reference", "J");

    JSObjectRef objRef = JSObjectMakeArray((JSContextRef)ctx, (size_t)len, (len==0)?NULL:elements,
                                           &exception);
    JSValueProtect((JSContextRef) ctx, objRef);

    env->SetLongField( out, fid, (jlong)objRef );

    fid = env->GetFieldID(ret , "exception", "J");
    env->SetLongField( out, fid, (long) exception);

    delete [] elements;
    return out;
}

JSC_NATIVE_METHOD(JSObject,jobject,NJSmakeDate) (JSC_NATIVE_PARAMS, jlong ctx, jlongArray args) {
    JSValueRef exception = NULL;

    int i;
    jsize len = env->GetArrayLength(args);
    jlong *values = env->GetLongArrayElements(args, 0);
    JSValueRef* elements = new JSValueRef[len];
    for (i=0; i<len; i++) {
        elements[i] = (JSValueRef) values[i];
    }
    env->ReleaseLongArrayElements(args, values, 0);

    jclass ret = env->FindClass("io/nodekit/engine/javascriptcore/JSValue$JNIReturnObject");
    jmethodID cid = env->GetMethodID(ret,"<init>","()V");
    jobject out = env->NewObject(ret, cid);

    jfieldID fid = env->GetFieldID(ret , "reference", "J");

    JSObjectRef objRef = JSObjectMakeDate((JSContextRef)ctx, (size_t)len, (len==0)?NULL:elements,
                                          &exception);
    JSValueProtect((JSContextRef) ctx, objRef);
    env->SetLongField( out, fid, (jlong) objRef );

    fid = env->GetFieldID(ret , "exception", "J");
    env->SetLongField( out, fid, (long) exception);

    delete [] elements;
    return out;
}

JSC_NATIVE_METHOD(JSObject,jobject,NJSmakeError) (JSC_NATIVE_PARAMS, jlong ctx, jlongArray args) {
    JSValueRef exception = NULL;

    int i;
    jsize len = env->GetArrayLength(args);
    jlong *values = env->GetLongArrayElements(args, 0);
    JSValueRef* elements = new JSValueRef[len];
    for (i=0; i<len; i++) {
        elements[i] = (JSValueRef) values[i];
    }
    env->ReleaseLongArrayElements(args, values, 0);

    jclass ret = env->FindClass("io/nodekit/engine/javascriptcore/JSValue$JNIReturnObject");
    jmethodID cid = env->GetMethodID(ret,"<init>","()V");
    jobject out = env->NewObject(ret, cid);

    jfieldID fid = env->GetFieldID(ret , "reference", "J");

    JSObjectRef objRef = JSObjectMakeError((JSContextRef)ctx, (size_t)len, (len==0)?NULL:elements,
                                           &exception);
    JSValueProtect((JSContextRef) ctx, objRef);
    env->SetLongField( out, fid, (long) objRef );

    fid = env->GetFieldID(ret , "exception", "J");
    env->SetLongField( out, fid, (long) exception);

    delete [] elements;
    return out;
}

JSC_NATIVE_METHOD(JSObject,jobject,NJSmakeRegExp) (JSC_NATIVE_PARAMS, jlong ctx, jlongArray args) {
    JSValueRef exception = NULL;

    int i;
    jsize len = env->GetArrayLength(args);
    jlong *values = env->GetLongArrayElements(args, 0);
    JSValueRef* elements = new JSValueRef[len];
    for (i=0; i<len; i++) {
        elements[i] = (JSValueRef) values[i];
    }
    env->ReleaseLongArrayElements(args, values, 0);

    jclass ret = env->FindClass("io/nodekit/engine/javascriptcore/JSValue$JNIReturnObject");
    jmethodID cid = env->GetMethodID(ret,"<init>","()V");
    jobject out = env->NewObject(ret, cid);

    jfieldID fid = env->GetFieldID(ret , "reference", "J");

    JSObjectRef objRef = JSObjectMakeRegExp((JSContextRef)ctx, (size_t)len, (len==0)?NULL:elements,
                                            &exception);
    JSValueProtect((JSContextRef) ctx, objRef);
    env->SetLongField( out, fid, (long) objRef );

    fid = env->GetFieldID(ret , "exception", "J");
    env->SetLongField( out, fid, (long) exception);

    delete [] elements;
    return out;
}

JSC_NATIVE_METHOD(JSObject,jobject,makeFunction) (JSC_NATIVE_PARAMS, jlong ctx, jlong name,
                                       jlongArray parameterNames, jlong body, jlong sourceURL, jint startingLineNumber) {

    JSValueRef exception = NULL;

    int i;
    jsize len = env->GetArrayLength(parameterNames);
    jlong *parameters = env->GetLongArrayElements(parameterNames, 0);
    JSStringRef* parameterNameArr = new JSStringRef[len];
    for (i=0; i<len; i++) {
        parameterNameArr[i] = (JSStringRef) parameters[i];
    }
    env->ReleaseLongArrayElements(parameterNames, parameters, 0);

    jclass ret = env->FindClass("io/nodekit/engine/javascriptcore/JSValue$JNIReturnObject");
    jmethodID cid = env->GetMethodID(ret,"<init>","()V");
    jobject out = env->NewObject(ret, cid);

    jfieldID fid = env->GetFieldID(ret , "reference", "J");

    JSObjectRef objref = JSObjectMakeFunction(
            (JSContextRef)ctx,
            (JSStringRef)name,
            (unsigned)len,
            (len==0)?NULL:parameterNameArr,
            (JSStringRef) body,
            (JSStringRef) sourceURL,
            (int)startingLineNumber,
            &exception);
    JSValueProtect((JSContextRef) ctx, objref);
    env->SetLongField( out, fid, (long)objref);

    fid = env->GetFieldID(ret , "exception", "J");
    env->SetLongField( out, fid, (long) exception);

    delete [] parameterNameArr;
    return out;
}

JSC_NATIVE_METHOD(JSObject,jlong,NJSgetPrototype) (JSC_NATIVE_PARAMS, jlong ctx, jlong object) {
    JSValueRef value = JSObjectGetPrototype((JSContextRef)ctx, (JSObjectRef)object);
    JSValueProtect((JSContextRef)ctx, value);
    return (long)value;
}

JSC_NATIVE_METHOD(JSObject,void,NJSsetPrototype) (JSC_NATIVE_PARAMS, jlong ctx, jlong object, jlong value) {
    JSObjectSetPrototype((JSContextRef)ctx, (JSObjectRef)object, (JSValueRef)value);
}

JSC_NATIVE_METHOD(JSObject,jboolean,NJShasProperty) (JSC_NATIVE_PARAMS, jlong ctx, jlong object, jlong propertyName) {
    return JSObjectHasProperty((JSContextRef)ctx, (JSObjectRef) object, (JSStringRef)propertyName);
}

JSC_NATIVE_METHOD(JSObject,jobject,NJSgetProperty) (JSC_NATIVE_PARAMS, jlong ctx, jlong object,
                                      jlong propertyName) {

    JSValueRef exception = NULL;

    jclass ret = env->FindClass("io/nodekit/engine/javascriptcore/JSValue$JNIReturnObject");
    jmethodID cid = env->GetMethodID(ret,"<init>","()V");
    jobject out = env->NewObject(ret, cid);

    jfieldID fid = env->GetFieldID(ret , "reference", "J");

    JSValueRef value = JSObjectGetProperty((JSContextRef)ctx, (JSObjectRef)object, (JSStringRef)propertyName,
                                           &exception);
    JSValueProtect((JSContextRef) ctx, value);

    env->SetLongField( out, fid, (long)value);

    fid = env->GetFieldID(ret , "exception", "J");
    env->SetLongField( out, fid, (long) exception);

    return out;
}

JSC_NATIVE_METHOD(JSObject,jobject,NJSsetProperty) (JSC_NATIVE_PARAMS, jlong ctx, jlong object, jlong propertyName,
                                      jlong value, jint attributes) {

    JSValueRef exception = NULL;

    jclass ret = env->FindClass("io/nodekit/engine/javascriptcore/JSValue$JNIReturnObject");
    jmethodID cid = env->GetMethodID(ret,"<init>","()V");
    jobject out = env->NewObject(ret, cid);

    JSObjectSetProperty((JSContextRef)ctx, (JSObjectRef) object, (JSStringRef) propertyName,
                        (JSValueRef)value, (JSPropertyAttributes)attributes, &exception);

    jfieldID fid = env->GetFieldID(ret , "exception", "J");
    env->SetLongField( out, fid, (long) exception);

    return out;
}

JSC_NATIVE_METHOD(JSObject,jobject,NJSdeleteProperty) (JSC_NATIVE_PARAMS, jlong ctx, jlong object,
                                         jlong propertyName) {

    JSValueRef exception = NULL;

    jclass ret = env->FindClass("io/nodekit/engine/javascriptcore/JSValue$JNIReturnObject");
    jmethodID cid = env->GetMethodID(ret,"<init>","()V");
    jobject out = env->NewObject(ret, cid);

    jfieldID fid = env->GetFieldID(ret ,"bool", "Z");

    bool bval = (bool) JSObjectDeleteProperty((JSContextRef)ctx, (JSObjectRef) object,
                                              (JSStringRef) propertyName, &exception);

    env->SetBooleanField( out, fid, bval);

    fid = env->GetFieldID(ret , "exception", "J");
    env->SetLongField( out, fid, (long) exception);

    return out;
}

JSC_NATIVE_METHOD(JSObject,jobject,NJSgetPropertyAtIndex) (JSC_NATIVE_PARAMS, jlong ctx, jlong object,
                                             jint propertyIndex) {

    JSValueRef exception = NULL;

    jclass ret = env->FindClass("io/nodekit/engine/javascriptcore/JSValue$JNIReturnObject");
    jmethodID cid = env->GetMethodID(ret,"<init>","()V");
    jobject out = env->NewObject(ret, cid);

    jfieldID fid = env->GetFieldID(ret , "reference", "J");

    JSValueRef value = JSObjectGetPropertyAtIndex((JSContextRef)ctx, (JSObjectRef) object,
                                                  (unsigned)propertyIndex, &exception);
    JSValueProtect((JSContextRef)ctx, value);

    env->SetLongField( out, fid, (long)value );

    fid = env->GetFieldID(ret , "exception", "J");
    env->SetLongField( out, fid, (long) exception);

    return out;
}

JSC_NATIVE_METHOD(JSObject,jobject,NJSsetPropertyAtIndex) (JSC_NATIVE_PARAMS, jlong ctx, jlong object,
                                             jint propertyIndex, jlong value) {

    JSValueRef exception = NULL;

    jclass ret = env->FindClass("io/nodekit/engine/javascriptcore/JSValue$JNIReturnObject");
    jmethodID cid = env->GetMethodID(ret,"<init>","()V");
    jobject out = env->NewObject(ret, cid);

    jfieldID fid = env->GetFieldID(ret , "reference", "J");

    JSObjectSetPropertyAtIndex((JSContextRef)ctx, (JSObjectRef) object, (unsigned) propertyIndex,
                               (JSValueRef)value, &exception);

    fid = env->GetFieldID(ret , "exception", "J");
    env->SetLongField( out, fid, (long) exception);


    return out;
}

JSC_NATIVE_METHOD(JSObject,jlong,NJSgetPrivate) (JSC_NATIVE_PARAMS, jlong object) {
    return (long) JSObjectGetPrivate((JSObjectRef) object);
}

JSC_NATIVE_METHOD(JSObject,jboolean,NJSsetPrivate) (JSC_NATIVE_PARAMS, jlong object, jlong data) {
    return JSObjectSetPrivate((JSObjectRef) object, (void*) data);
}

JSC_NATIVE_METHOD(JSObject,jboolean,NJSisFunction) (JSC_NATIVE_PARAMS, jlong ctx, jlong object) {
    return (jboolean) JSObjectIsFunction((JSContextRef)ctx, (JSObjectRef) object);
}

JSC_NATIVE_METHOD(JSObject,jobject,NJScallAsFunction) (JSC_NATIVE_PARAMS, jlong ctx, jlong object,
                                         jlong thisObject, jlongArray args) {
    JSValueRef exception = NULL;

    int i;
    jsize len = env->GetArrayLength(args);
    jlong *values = env->GetLongArrayElements(args, 0);
    JSValueRef* elements = new JSValueRef[len];
    for (i=0; i<len; i++) {
        elements[i] = (JSValueRef) values[i];
    }
    env->ReleaseLongArrayElements(args, values, 0);

    jclass ret = env->FindClass("io/nodekit/engine/javascriptcore/JSValue$JNIReturnObject");
    jmethodID cid = env->GetMethodID(ret,"<init>","()V");
    jobject out = env->NewObject(ret, cid);

    jfieldID fid = env->GetFieldID(ret , "reference", "J");

    JSValueRef value = JSObjectCallAsFunction((JSContextRef)ctx, (JSObjectRef) object, (JSObjectRef) thisObject,
                                              (size_t)len, (len==0)?NULL:elements, &exception);
    JSValueProtect((JSContextRef) ctx, value);

    env->SetLongField( out, fid, (long)value);

    fid = env->GetFieldID(ret , "exception", "J");
    env->SetLongField( out, fid, (long) exception);

    delete [] elements;
    return out;
}

JSC_NATIVE_METHOD(JSObject,jboolean,NJSisConstructor) (JSC_NATIVE_PARAMS, jlong ctx, jlong object) {
    return (jboolean) JSObjectIsConstructor((JSContextRef)ctx, (JSObjectRef)object);
}

JSC_NATIVE_METHOD(JSObject,jobject,NJScallAsConstructor) (JSC_NATIVE_PARAMS, jlong ctx, jlong object,
                                            jlongArray args) {

    JSValueRef exception = NULL;

    int i;
    jsize len = env->GetArrayLength(args);
    jlong *values = env->GetLongArrayElements(args, 0);
    JSValueRef* elements = new JSValueRef[len];
    for (i=0; i<len; i++) {
        elements[i] = (JSValueRef) values[i];
    }
    env->ReleaseLongArrayElements(args, values, 0);

    jclass ret = env->FindClass("io/nodekit/engine/javascriptcore/JSValue$JNIReturnObject");
    jmethodID cid = env->GetMethodID(ret,"<init>","()V");
    jobject out = env->NewObject(ret, cid);

    jfieldID fid = env->GetFieldID(ret , "reference", "J");

    JSValueRef value = JSObjectCallAsConstructor((JSContextRef)ctx, (JSObjectRef) object,
                                                 (size_t)len, (len==0)?NULL:elements, &exception);
    JSValueProtect((JSContextRef) ctx, value);

    env->SetLongField( out, fid, (long)value);

    fid = env->GetFieldID(ret , "exception", "J");
    env->SetLongField( out, fid, (long) exception);

    delete [] elements;
    return out;
}

JSC_NATIVE_METHOD(JSObject,jlong,NJScopyPropertyNames) (JSC_NATIVE_PARAMS, jlong ctx, jlong object) {
    JSPropertyNameArrayRef ref = JSObjectCopyPropertyNames((JSContextRef)ctx, (JSObjectRef)object);
    JSPropertyNameArrayRetain(ref);
    return (long)ref;
}

JSC_NATIVE_METHOD(JSObject,jlongArray,NJSgetPropertyNames) (JSC_NATIVE_PARAMS,jlong propertyNameArray) {
    size_t count = JSPropertyNameArrayGetCount((JSPropertyNameArrayRef)propertyNameArray);
    jlongArray retArray = env->NewLongArray(count);
    jlong* stringRefs = new jlong[count];
    for (size_t i=0; i<count; i++) {
        stringRefs[i] = (long) JSStringRetain(JSPropertyNameArrayGetNameAtIndex(
                (JSPropertyNameArrayRef)propertyNameArray, i));
    }
    env->SetLongArrayRegion(retArray,0,count,stringRefs);

    return retArray;
}

JSC_NATIVE_METHOD(JSObject,void,NJSreleasePropertyNames) (JSC_NATIVE_PARAMS, jlong propertyNameArray) {
    JSPropertyNameArrayRelease((JSPropertyNameArrayRef)propertyNameArray);
}

File: engine-jsc/src/main/java/io/nodekit/engine/javascriptcore/JSObject.java
package io.nodekit.engine.javascriptcore;

import io.nodekit.nkscripting.NKScriptValue;
import io.nodekit.engine.javascriptcore.types.JSObjectPropertiesMap;
import io.nodekit.engine.javascriptcore.util.JSException;

import android.webkit.ValueCallback;
import android.support.annotation.Nullable;

import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * A JavaScript object.
 */
public class JSObject extends JSValue implements NKScriptValue {

    //
    // NKScriptValue interface
    //

    public void invokeMethod(String method, Object[] arguments, ValueCallback<String> completionHandler) {
        JSValue result = property(method).jsvalueToFunction().apply(null, arguments);
        completionHandler.onReceiveValue(result.toString());
    }

    public void invokeMethod(String method, Object[] arguments) {
        property(method).jsvalueToFunction().apply(null, arguments);
    }

    public void defineProperty(String key, Object value) {
        property(key, value);
    }

    public void deleteProperty(String key) {
        jsobjectDeleteProperty(key);
    }

    public void hasProperty(String key, ValueCallback<String> completionHandler) {
        boolean result = jsobjectHasProperty(key);
        completionHandler.onReceiveValue(result ? "true" : null);
    }

    public void valueForProperty(String property, ValueCallback<String> completionHandler) {
        JSValue result = property(property);
        completionHandler.onReceiveValue(result.toString());
    }

    public  void setValue(Object value, String forProperty) {
        property(forProperty, value);
    }


    protected JSFunction isInstanceOf = null;

    private JSObject thiz = null;

    protected final List<JSObject> zombies = new ArrayList<>();

    /**
     * Specifies that a property has no special attributes.
     */
    public static int JSPropertyAttributeNone = 0;
    /**
     * Specifies that a property is read-only.
     */
    public static int JSPropertyAttributeReadOnly = 1 << 1;
    /**
     * Specifies that a property should not be enumerated by
     * JSPropertyEnumerators and JavaScript for...in loops.
     */
    public static int JSPropertyAttributeDontEnum = 1 << 2;
    /**
     * Specifies that the delete operation should fail on a property.
     */
    public static int JSPropertyAttributeDontDelete = 1 << 3;

    /**
     * Creates a new, empty JavaScript object.  In JS:
     * <pre>
     * {@code
     * var obj = {}; // OR
     * var obj = new Object();
     * }
     * </pre>
     *
     * @param ctx The JSContext to create the object in
     *
     */
    public JSObject(JSContext ctx) {
        context = ctx;
        context.sync(new Runnable() {
            @Override
            public void run() {
                valueRef = NJSmake(context.contextRef(), 0L);
            }
        });
        context.persistObject(this);
    }

    /**
     * Called only by convenience subclasses.  If you use
     * this, you must set context and valueRef yourself.
     */
    public JSObject() {
    }

    /**
     * Wraps an existing object from JavaScript
     *
     * @param objRef The JavaScriptCore object reference
     * @param ctx    The JSContext of the reference
     *
     */
    protected JSObject(final long objRef, JSContext ctx) {
        super(objRef, ctx);
        context.persistObject(this);
    }

    /**
     * Creates a new object with function properties set for each method
     * in the defined interface.
     * In JS:
     * <pre>
     * {@code
     * var obj = {
     *     func1: function(a)   { alert(a); },
     *     func2: function(b,c) { alert(b+c); }
     * };
     * }
     * </pre>
     * Where func1, func2, etc. are defined in interface 'iface'.  This JSObject
     * must implement 'iface'.
     *
     * @param ctx   The JSContext to create the object in
     * @param iface The Java Interface defining the methods to expose to JavaScript
     *
     */
    public JSObject(JSContext ctx, final Class<?> iface) {
        context = ctx;
        context.sync(new Runnable() {
            @Override
            public void run() {
                valueRef = NJSmake(context.contextRef(), 0L);
                Method[] methods = iface.getDeclaredMethods();
                for (Method m : methods) {
                    JSObject f = new JSFunction(context, m,
                            JSObject.class, JSObject.this);
                    property(m.getName(), f);
                }
            }
        });
        context.persistObject(this);
    }

    /**
     * Creates a new function object with the entries in 'map' set as properties.
     *
     * @param ctx The JSContext to create object in
     * @param map The map containing the properties
     */
    @SuppressWarnings("unchecked")
    public JSObject(JSContext ctx, final Map map) {
        this(ctx);
        new JSObjectPropertiesMap<>(this, Object.class).putAll(map);
    }

    /**
     * Determines if the object contains a given property
     *
     * @param prop The property to test the existence of
     * @return true if the property exists on the object, false otherwise
     *
     */
    public boolean jsobjectHasProperty(final String prop) {
        JNIReturnClass runnable = new JNIReturnClass() {
            @Override
            public void run() {
                jni = new JNIReturnObject();
                jni.bool = NJShasProperty(context.contextRef(), valueRef, new JSString(prop).stringRef());
            }
        };
        context.sync(runnable);
        return runnable.jni.bool;
    }

    /**
     * Gets the property named 'prop'
     *
     * @param prop The name of the property to fetch
     * @return The JSValue of the property, or null if it does not exist
     *
     */
    public JSValue property(final String prop) {
        JNIReturnClass runnable = new JNIReturnClass() {
            @Override
            public void run() {
                jni = NJSgetProperty(context.contextRef(), valueRef, new JSString(prop).stringRef());
            }
        };
        context.sync(runnable);
        if (runnable.jni.exception != 0) {
            context.throwJSException(new JSException(new JSValue(runnable.jni.exception, context)));
            return new JSValue(context);
        }
        return new JSValue(runnable.jni.reference, context);
    }

    /**
     * Sets the value of property 'prop'
     *
     * @param prop       The name of the property to set
     * @param value      The Java object to set.  The Java object will be converted to a JavaScript object
     *                   automatically.
     * @param attributes And OR'd list of JSProperty constants
     *
     */
    public void property(final String prop, final Object value, final int attributes) {
        JNIReturnClass runnable = new JNIReturnClass() {
            @Override
            public void run() {
                JSString name = new JSString(prop);
                jni = NJSsetProperty(
                        context.contextRef(),
                        valueRef,
                        name.stringRef,
                        (value instanceof JSValue) ? ((JSValue) value).valueRef() : new JSValue(context, value).valueRef(),
                        attributes);
            }
        };
        context.sync(runnable);
        if (runnable.jni.exception != 0) {
            context.throwJSException(new JSException(new JSValue(runnable.jni.exception, context)));
        }
    }

    /**
     * Sets the value of property 'prop'.  No JSProperty attributes are set.
     *
     * @param prop  The name of the property to set
     * @param value The Java object to set.  The Java object will be converted to a JavaScript object
     *              automatically.
     *
     */
    public void property(String prop, Object value) {
        property(prop, value, JSPropertyAttributeNone);
    }

    /**
     * Deletes a property from the object
     *
     * @param prop The name of the property to delete
     * @return true if the property was deleted, false otherwise
     *
     */
    public boolean jsobjectDeleteProperty(final String prop) {
        JNIReturnClass runnable = new JNIReturnClass() {
            @Override
            public void run() {
                JSString name = new JSString(prop);
                jni = NJSdeleteProperty(context.contextRef(), valueRef, name.stringRef());
            }
        };
        context.sync(runnable);
        if (runnable.jni.exception != 0) {
            context.throwJSException(new JSException(new JSValue(runnable.jni.exception, context)));
            return false;
        }
        return runnable.jni.bool;
    }

    /**
     * Returns the property at index 'index'.  Used for arrays.
     *
     * @param index The index of the property
     * @return The JSValue of the property at index 'index'
     *
     */
    public JSValue propertyAtIndex(final int index) {
        JNIReturnClass runnable = new JNIReturnClass() {
            @Override
            public void run() {
                jni = NJSgetPropertyAtIndex(context.contextRef(), valueRef, index);
            }
        };
        context.sync(runnable);
        if (runnable.jni.exception != 0) {
            context.throwJSException(new JSException(new JSValue(runnable.jni.exception, context)));
            return new JSValue(context);
        }
        return new JSValue(runnable.jni.reference, context);
    }

    /**
     * Sets the property at index 'index'.  Used for arrays.
     *
     * @param index The index of the property to set
     * @param value The Java object to set, will be automatically converted to a JavaScript value
     *
     */
    public void propertyAtIndex(final int index, final Object value) {
        JNIReturnClass runnable = new JNIReturnClass() {
            @Override
            public void run() {
                jni = NJSsetPropertyAtIndex(context.contextRef(), valueRef, index,
                        (value instanceof JSValue) ? ((JSValue) value).valueRef() : new JSValue(context, value).valueRef());
            }
        };
        context.sync(runnable);
        if (runnable.jni.exception != 0) {
            context.throwJSException(new JSException(new JSValue(runnable.jni.exception, context)));
        }
    }

    private abstract class StringArrayReturnClass implements Runnable {
        public String[] sArray;
    }

    /**
     * Gets the list of set property names on the object
     *
     * @return A string array containing the property names
     *
     */
    public String[] propertyNames() {
        StringArrayReturnClass runnable = new StringArrayReturnClass() {
            @Override
            public void run() {
                long propertyNameArray = NJScopyPropertyNames(context.contextRef(), valueRef);
                long[] refs = NJSgetPropertyNames(propertyNameArray);
                String[] names = new String[refs.length];
                for (int i = 0; i < refs.length; i++) {
                    JSString name = new JSString(refs[i]);
                    names[i] = name.toString();
                }
                NJSreleasePropertyNames(propertyNameArray);
                sArray = names;
            }
        };
        context.sync(runnable);
        return runnable.sArray;
    }

    /**
     * Determines if the object is a function
     *
     * @return true if the object is a function, false otherwise
     *
     */
    public boolean jsvalueOrObjectIsFunction() {
        JNIReturnClass runnable = new JNIReturnClass() {
            @Override
            public void run() {
                jni = new JNIReturnObject();
                jni.bool = NJSisFunction(context.contextRef(), valueRef);
            }
        };
        context.sync(runnable);
        return runnable.jni.bool;
    }

    /**
     * Determines if the object is a constructor
     *
     * @return true if the object is a constructor, false otherwise
     *
     */
    public boolean jsobjectIsConstructor() {
        JNIReturnClass runnable = new JNIReturnClass() {
            @Override
            public void run() {
                jni = new JNIReturnObject();
                jni.bool = NJSisConstructor(context.contextRef(), valueRef);
            }
        };
        context.sync(runnable);
        return runnable.jni.bool;
    }

    @Override
    public int hashCode() {
        return valueRef().intValue();
    }

    @Override
    protected void finalize() throws Throwable {
        super.finalize();
        context.finalizeObject(this);
    }

    protected void setThis(JSObject thiz) {
        this.thiz = thiz;
    }

    public JSObject getThis() {
        return thiz;
    }

    public JSValue __nullFunc() {
        return new JSValue(context);
    }

        //
    // JS Function
    //
  
    public void callWithArguments(Object[] arguments, @Nullable ValueCallback<String> completionHandler) {
        // overriden by JSFunction 
    }


    //
    // JS Array
    //

    public void valueAtIndex(int index, ValueCallback<String> completionHandler) {
    }

    public  void setValue(Object value, int atIndex) {
           // overriden by JSArray 
    }

    /* Native Methods */

    protected native long NJSmake(long ctx, long data);

    protected native long NJSmakeInstance(long ctx);

    protected native JNIReturnObject NJSmakeArray(long ctx, long[] args);

    protected native JNIReturnObject NJSmakeDate(long ctx, long[] args);

    protected native JNIReturnObject NJSmakeError(long ctx, long[] args);

    protected native JNIReturnObject NJSmakeRegExp(long ctx, long[] args);

    protected native long NJSgetPrototype(long ctx, long object);

    protected native void NJSsetPrototype(long ctx, long object, long value);

    protected native boolean NJShasProperty(long ctx, long object, long propertyName);

    protected native JNIReturnObject NJSgetProperty(long ctx, long object, long propertyName);

    protected native JNIReturnObject NJSsetProperty(long ctx, long object, long propertyName, long value, int attributes);

    protected native JNIReturnObject NJSdeleteProperty(long ctx, long object, long propertyName);

    protected native JNIReturnObject NJSgetPropertyAtIndex(long ctx, long object, int propertyIndex);

    protected native JNIReturnObject NJSsetPropertyAtIndex(long ctx, long object, int propertyIndex, long value);

    protected native long NJSgetPrivate(long object);

    protected native boolean NJSetPrivate(long object, long data);

    protected native boolean NJSisFunction(long ctx, long object);

    protected native JNIReturnObject NJScallAsFunction(long ctx, long object, long thisObject, long[] args);

    protected native boolean NJSisConstructor(long ctx, long object);

    protected native JNIReturnObject NJScallAsConstructor(long ctx, long object, long[] args);

    protected native long NJScopyPropertyNames(long ctx, long object);

    protected native long[] NJSgetPropertyNames(long propertyNameArray);

    protected native void NJSreleasePropertyNames(long propertyNameArray);

    protected native long NJSmakeFunctionWithCallback(long ctx, long name);

    protected native JNIReturnObject NJSmakeFunction(long ctx, long name, long[] parameterNames,
                                                  long body, long sourceURL, int startingLineNumber);

    private abstract class JNIReturnClass implements Runnable {
        JNIReturnObject jni;
    }
}